### ISSUE TITLE
Numpy operations on IREE DeviceArray python bindings are unexpectedly expensive when using Vulkan

### ISSUE THREAD

**Issue by qedawkins** [iree-org/iree#11430]
I've been running into difficulties with operations being unexpectedly expensive on the results of a model when using the IREE python bindings.

## Example script

Starting with the following example script, I'm taking the result of a model that fills a tensor with `1.0` and then try to scale the result in a few different ways.

```python
import iree.runtime as ireert
from iree.runtime import get_driver, get_device
import iree.compiler as ireec
import numpy as np
import time

module = """ 
builtin.module {
  func.func @forward() -> tensor<1x3x512x512xf16> {
    %cst = arith.constant 1.000000e+00 : f16
    %0 = tensor.empty() : tensor<1x3x512x512xf16>
    %1 = linalg.fill ins(%cst : f16) outs(%0 : tensor<1x3x512x512xf16>) -> tensor<1x3x512x512xf16>
    return %1 : tensor<1x3x512x512xf16>
  }
}
"""

device = "vulkan"

flatbuffer = ireec.compile_str(
    module,
    target_backends=[device],
)

config = ireert.Config(driver_name=device)
vm_module = ireert.VmModule.from_flatbuffer(config.vm_instance, flatbuffer)
ctx = ireert.SystemContext(config=config)
ctx.add_vm_module(vm_module)
forward = ctx.modules.module["forward"]

start = time.time()
result = forward()
print(result)
print("forward: ", (time.time() - start) * 1000);

# numpy scaling
start = time.time()
np_result = result.to_host()
scaled = 255.0 * np_result
print("numpy scaling: ", (time.time() - start) * 1000)

# numpy scaling with copy
start = time.time()
np_result = np.copy(result.to_host())
copy_end = time.time()
scaled = 255.0 * np_result
print("copy cost: ", (copy_end - start) * 1000)
print("copy + numpy scaling: ", (time.time() - start) * 1000)

# torch scaling without copy
import torch
start = time.time()
np_result = torch.from_numpy(result.to_host())
scaled = (255.0 * np_result).numpy()
print("torch scaling time: ", (time.time() - start) * 1000)

```

## Results

If I run this script I see something like the following timings
```
<IREE DeviceArray: shape=[1, 3, 512, 512], dtype=float16>
forward:  0.4475116729736328
numpy scaling:  14847.1519947052
copy cost:  1061.7661476135254
copy + numpy scaling:  1066.7922496795654
torch scaling time:  11.160612106323242
```

Notably, when doing `result.to_host()` and scaling the resulting numpy array it takes over 14 seconds to do. If I instead first call `np.copy` and then do the same scaling the copy takes over a second but then the operation on the array only takes a few more milliseconds. Finally, if I convert the result to a PyTorch tensor and then scale and convert back to a numpy array the overall process takes ~11ms.

While my focus is for f16, if I do the same for f32 the results look a bit different.
```
<IREE DeviceArray: shape=[1, 3, 512, 512], dtype=float32>
forward:  0.5207061767578125
numpy scaling:  1431.398630142212
copy cost:  2265.2816772460938
copy + numpy scaling:  2266.0908699035645
torch scaling time:  12.434720993041992
```

Namely the cost of scaling the result directly is a lot cheaper but the cost of copying the array is still high.

## Setup info

Most recent IREE commit: dc15a278e838
pytorch version: `1.14.0.dev20221114+cpu`

**Comment by benvanik** [comment]
(14 seconds is a long time but doing a to_host is one of the worst things that can be present in performance-sensitive code - think of to_host/to_device like blocking TCP socket send/recv - an application wanting to do lots of transfers will want to issue command buffers with transfer operations that provide for batching and allocation control)

**Comment by qedawkins** [comment]
> (14 seconds is a long time but doing a to_host is one of the worst things that can be present in performance-sensitive code - think of to_host/to_device like blocking TCP socket send/recv - an application wanting to do lots of transfers will want to issue command buffers with transfer operations that provide for batching and allocation control)

Makes sense. I assumed I would have to eventually send to host though as we are converting the output to a jpg, but the scaling could happen before transferring. Regardless the 14 seconds and 1s copy was mainly what I was having troubles with.

**Comment by benvanik** [comment]
Yeah if you can do the scaling in the model you'll get the best performance as we can fuse that with whatever is producing the output and vectorize/etc.

I suspect part of to_host is the implicit allocation that needs to be done (host array needs to be allocated) - a tracy/perf run would help identify what's causing it. May be something silly and if we can rule out Vulkan API overhead that'd help!

**Comment by ScottTodd** [comment]
There are also a few methods in https://github.com/iree-org/iree/blob/main/runtime/bindings/python/iree/runtime/array_interop.py that are implemented inefficiently (e.g. metadata only operations that synchronize with the host). I'm not sure how relevant that is for this case though.

**Comment by qedawkins** [comment]
For context the code in SHARK used to look something like this
```
result = np.copy(np.asarray(results, dtype=results.dtype)
```
which was taking ~1s to copy on Windows. I ended up at `array_interop.py` and assumed this was ending up transferring to host.

**Comment by qedawkins** [comment]
> I suspect part of to_host is the implicit allocation that needs to be done (host array needs to be allocated) - a tracy/perf run would help identify what's causing it. May be something silly and if we can rule out Vulkan API overhead that'd help!

Can I get a tracy profile on stuff that happens after the model in python?

**Comment by jpienaar** [comment]
Is getting the tracy profile from the run the pending item here?

**Comment by qedawkins** [comment]
Sorry should have updated this issue. I did get the tracy profile but it didn't show anything for the regions I was concerned with. The 14 seconds to do that scaling is still unclear to me but it's also not clear that it is an issue with IREE. For the particular case where I came across this, the solution to add the scaling inside the model worked fine, although that just avoids the problem described here.

As a result I'm fine if this issue is closed unless someone else thinks this is clearly a problem with IREE.
